gensim/models/indexer.py

```diff
@@ -0,0 +1,246 @@
+# Licensed under the GNU LGPL v2.1 - http://www.gnu.org/licenses/lgpl.html
+
+
+"""
+Using 3rd party k-NN libraries to improve performance for method DocvecsArray.most_similar().
+More information can be found at: https://github.com/piskvorky/gensim/wiki/Ideas-&-Feature-proposals#integrate-a-fast-k-nn-library
+"""
+
+from doc2vec import Doc2Vec
+from word2vec import Word2Vec
+from annoy import AnnoyIndex
+
+class Indexer(object):
+    """
+    Base class for indexers
+    """
+
+    def __init__(self, model=None, **kwargs):
+        """
+        Create a new indexer
+
+        Args:
+            model:   Instance of Doc2Vec or Word2Vec
+            kwargs:  Additional named parameters
+        """
+        if model is None:
+            raise Exception("Invalid model parameter. Please provide model with an instance of Doc2Vec or Word2Vec")
+        if type(model) is Doc2Vec:
+            self._init_doc2vec_(model)
+        elif type(model) is Word2Vec:
+            self._init_word2vec_(model)
+        else:
+            raise Exception("Invalid model parameter. Please provide model with an instance of Doc2Vec or Word2Vec")
+        self.start_indexing()
+
+    def _init_doc2vec_(self, model):
+        docvecs = model.docvecs
+        docvecs.init_sims()
+        size = len(docvecs.doctag_syn0norm)
+
+        for i in range(size):
+            self.add_item(docvecs.offset2doctag[i], docvecs.doctag_syn0norm[i])
+
+    def _init_word2vec_(self, model):
+        model.init_sims()
+        size = len(model.syn0norm)
+
+        for i in range(size):
+            self.add_item(model.index2word[i], model.syn0norm[i])
+
+    def get_item(self, label):
+        """
+        Get an item's vector by its label
+
+        Args:
+            label: The label
+
+        Returns:
+            The item's vector if have. Otherwise, returns None
+        """
+        pass
+
+    def add_item(self, label, vector):
+        """
+        Add an item to index
+
+        Args:
+            label: the label of the item (must be unique).
+            vector: the item's vector.
+        """
+        pass
+
+    def start_indexing(self):
+        """
+        Start the indexing operation.
+        This method is supposed to run after all items are added
+        It may take time to finish.
+        """
+        pass
+
+    def get_nearest_items(self, vector, top_n=10):
+        """
+        Get nearest items for an item described by its vector
+
+        Args:
+            vector: The vector
+            top_n: Number of nearest items to get
+
+        Returns:
+            A list of tuples (label, similarity) for the nearest ones
+        """
+        pass
+
+    def save(self, file_name):
+        """
+        Dump internal data into disk
+
+        Args:
+            file_name:   Output file path
+
+        Returns:
+            Returns true on success. Otherwise, returns false
+        """
+        pass
+
+    def load(self, file_name):
```

Comment:
If `load` becomes a class method, there are a few issues that crop up:
* The initialization/indexing code based on the model is not needed when loading the index from disk.
* The `index_label_dict` and `label_index_dict` would need to saved/loaded in order to maintain the mapping between labels, ids, and vectors.

It seems like the process of building an index from a model could go in a class method, and `__init__` could take the index instead of the model. Not sure what to do about persisting the mappings, though.